cmd/src/debug_common.go

```diff
@@ -0,0 +1,109 @@
+package main
+
+import (
+	"archive/zip"
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/sourcegraph/src-cli/internal/exec"
+)
+
+/*
+General Stuff
+TODO: file issue on the existence of OAuth signKey which needs to be redacted
+TODO: Create getSiteConfig function
+*/
+
+type archiveFile struct {
+	name string
+	data []byte
+	err  error
+}
+
+func archiveFileFromCommand(ctx context.Context, path, cmd string, args ...string) *archiveFile {
+	f := &archiveFile{name: path}
+	f.data, f.err = exec.CommandContext(ctx, cmd, args...).CombinedOutput()
+	if f.err != nil {
+		f.err = errors.Wrapf(f.err, "executing command: %s %s: received error: %s", cmd, strings.Join(args, " "), f.data)
+	}
+	return f
+}
+
+// This function prompts the user to confirm they want to run the command
+func verify(confirmationText string) (bool, error) {
+	input := ""
+	for strings.ToLower(input) != "y" && strings.ToLower(input) != "n" {
+		fmt.Printf("%s [y/N]: ", confirmationText)
+		if _, err := fmt.Scanln(&input); err != nil {
+			return false, err
+		}
+	}
+
+	return strings.ToLower(input) == "y", nil
+}
+
+// setOpenFileLimits increases the limit of open files to the given number. This is needed
+// when doings lots of concurrent network requests which establish open sockets.
+func setOpenFileLimits(n uint64) error {
+
+	var rlimit syscall.Rlimit
+	err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &rlimit)
+	if err != nil {
+		return err
+	}
+
+	rlimit.Max = n
+	rlimit.Cur = n
+
+	return syscall.Setrlimit(syscall.RLIMIT_NOFILE, &rlimit)
+}
+
+// setupDebug takes the name of a base directory and returns the file pipe, zip writer,
+// and context needed for later archive functions. Don't forget to defer close on these
+// after calling setupDebug!
+func setupDebug(base string) (*os.File, *zip.Writer, context.Context, error) {
```

Comment:
Opinion:

- It's strange to return `ctx` here when all you do is create it (you don't require it to create the files, which is what the _meat_ of this function is doing)
- You are necessarily returning resources that need to be handled in the caller. This makes it a bit hard to see in one location what resources are actually being created and how they're handled. It would be more idiomatic to open the file and create the zip writer in each of the debug_k8s/_compose/_serv functions. This won't be so much code if you no longer need the open file limits modification.

The reason for the second point is that the current code separates the construction and cleanup, where in Go they're generally paired together (defer doesn't work across function boundaries).

Reply:
refactored to be more idiomatic ✅, thank you for the explanation